### ISSUE TITLE
banana decodes integer values greater than 2147483647 and smaller than sys.maxint as long on all platforms

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4169]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| toidinamai reported|
|-|-|
|Trac ID|trac#4169|
|Type|enhancement|
|Created|2009-12-14 14:00:39Z|
|Branch|https://github.com/twisted/twisted/tree/banana-more-ints-4169|

Hi,

I am working on a 64bit machine and use the Perspective Broker to wrap an API
that isn't really safe to call in my main python process.  This API does not
accept long objects and it throws an exception even when the long objects are
representable as ints, i.e. -sys.maxint-1 <= x <= sys.maxint.

Sometimes the integer values I send are greater than 2147483647 and on the
remote side they are decoded as longs.  This means that I cannot call the
wrapped API in those cases.  Unfortunately the values I send are themselves
contained in Python objects that mirror C structs and but it is nearly
impossible for me to walk the objects and downconvert the values myself.

I think when sending values through PB it should make sure they come out as the
same type they went in.  I tracked this down to banana and I can see that this
creates a problem because banana obviously wants to be platform independent.
Even though, I propose that banana should return values in the most appropriate
type on each platform, i.e. ints when they fit into python ints and longs
otherwise.  Just calling int wherever long is now called directly should fix
this an any version later than Python 2.2.

Regards,
Frank Benkstein.

Attachments:

* [long.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b7/7b74575fecc579bbcb7cfa407e84554b069d576e/8e0170ea0aed3c7c7128c77b46710dda0064c6ca.patch) (1614 bytes) - added by toidinamai on 2009-12-14 14:16:06Z - attempt to correct the problem
* [long-neg.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b7/7b74575fecc579bbcb7cfa407e84554b069d576e/c064cd190021b88dc9be80ec31e1bd635d402c04.patch) (2025 bytes) - added by toidinamai on 2009-12-28 07:46:11Z - add test-case for negative longs

<details><summary>Searchable metadata</summary>

```
trac-id__4169 4169
type__enhancement enhancement
reporter__toidinamai toidinamai
priority__normal normal
milestone__ 
branch__branches_banana_more_ints_4169 branches/banana-more-ints-4169
branch_author__toidinamai__exarkun toidinamai, exarkun
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__pb__spread__banana_ pb, spread, banana,
time__1260799239000000 1260799239000000
changetime__1263932076000000 1263932076000000
version__None None
owner__ 
cc__thijs cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra removed owner|
|-|-|

Thanks toidinamai, putting the patch up for review as described on the TwistedDevelopment page.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27781]) Branching to 'banana-more-ints-4169'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [27782]) Apply part of long.patch

refs [#4169](https://github.com/twisted/twisted/issues/4169)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to toidinamai|
|-|-|

I agree that banana should let values be represented by the "smallest" type available on the platform (which isn't the same as making the values that come out be of the same type as they went in - you could send a big int from a 64 bit platform to a 32 bit platform where it will have to come out as a long).

I applied part of the patch to the branch above: the unit test change (slightly modified to use `assertIsInstance`) and only change banana.py enough to make it pass.  Looks like there should be some more tests to cover the other cases.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

I'm pretty sure that the patch doesn't break this, but for compatibility with other languages' PB implementations the *wire* format should differentiate between regular and long ints based on a fixed size. Changing the Python-side implementation is less important considering how interchangeable 'long' and 'int' are.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

+1 to what itamar said.  We should leave the wire protocol how it is.

**Comment by twisted-trac** [comment]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|toidinamai commented|
|-|-|

I modified the test slightly to include negative longs.  I hope you can
also accept that removing the call to int in the "typebyte == INT" case
is also trivially correct because b1282int calls int in all code paths.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess commented|
|-|-|

(In [27899]) Apply the test_banana.py part of toidinamai's long-neg.patch and
update copyright dates.

refs [#4169](https://github.com/twisted/twisted/issues/4169)

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|TimAllen set owner to @exarkun|
|-|-|

This patch looks good; but it seems in comment:3 exarkun didn't apply all of
it to the branch. In particular, the patch removes all the casts from the
integer-handling parts of `Banana.dataReceived`, but the branch only removes
the cast from the LONGNUM case. If the cast isn't removed from LONGNEG, I would expect this branch's test to fail on 64-bit machines. Also, if `b1282int()` is always casting its result to int then the INT case doesn't need to re-cast it.

Assigning back to exarkun to add the rest of the patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

I applied as much of the patch as was necessary to make the unit tests pass.  If more changes are necessary, and the existing tests pass, then more tests are necessary.  iirc, all the tests passed on all platforms last time I touched things.

**Comment by twisted-trac** [comment]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|toidinamai commented|
|-|-|

Replying to [exarkun](#note_12):
My long-neg.patch adds the tests that make the other changes neccessary.
jesstess applied test part of it but not the actual code part.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to toidinamai|
|-|-|

Indeed, the newest version of the test does fail on a 64 bit platform.  Removing the `long` from the `LONGNEG` case does fix it.

It also seems quite clear that removing the `int` from the `INT` case is the correct thing to do.  However, that case appears to be untested: instead of deleting the `int`, I replaced it with `long`.  It seems like that should have caused something to fail, if the test coverage were complete.

Can you look into that?

**Comment by twisted-trac** [comment]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|toidinamai commented|
|-|-|

Replying to [exarkun](#note_14):
> It also seems quite clear that removing the `int` from the `INT` case is the correct thing to do.  However, that case appears to be untested: instead of deleting the `int`, I replaced it with `long`.  It seems like that should have caused something to fail, if the test coverage were complete.
>
> Can you look into that?

Just to be sure we are talking about the same thing the lines we are referring to are the following:

```
elif typebyte == INT:
    buffer = rest
    num = b1282int(num)
    gotItem(int(num))
```

The only non-exception control path in b1282int ends with

```
    return int(i)
```

I can't think of a test case that succeeds on int(i) but fails on int(int(i)).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

I'm looking for a test that fails on `long(int(i))`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

To elaborate somewhat, I don't actually think anyone might accidentally insert a `long` into the `typebyte == INT` case in this code.  However, the fact that when I do so intentionally none of the existing tests fail tells me that none of the existing tests is asserting anything about the type of the value passed to `gotItem` in this case.  If that's something that we care about (as it appears to be), then it should be tested.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

Regarding the two remaining fixes from toidinamai's original patch, I checked in one of them in r27967 and the other, together with the improved test coverage I was after, in r27968.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|TimAllen set owner to @exarkun|
|-|-|

Looks good to me. +1 for merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [28030]) Merge banana-more-ints-4169 

Author: toidinamai, exarkun
Reviewer: exarkun, TimAllen
Fixes: [#4169](https://github.com/twisted/twisted/issues/4169)

When decoding integers in `twisted.spread.banana`, try to represent the value as
an instance of `int` if it will fit.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
